src/integ_test/java/games/strategy/engine/lobby/server/login/LobbyLoginValidatorTest.java

```diff
@@ -23,98 +27,182 @@
-        name, mac, address));
+    final Map<String, String> response = new HashMap<>();
+    response.put(LobbyLoginValidator.ANONYMOUS_LOGIN, Boolean.TRUE.toString());
+    assertNull(generateChallenge(null).apply(challenge -> response));
+
     // create a user, verify we can't login with a username that already exists
-    new DbUserController().createUser(
-        new DBUser(
-            new DBUser.UserName(name),
-            new DBUser.UserEmail("none@none.none")),
-          new HashedPassword(MD5Crypt.crypt("foo")));
     // we should not be able to login now
-    assertNotNull(new LobbyLoginValidator().verifyConnection(validator.getChallengeProperties(name, address),
-        properties, name, mac, address));
+    assertNotNull(generateChallenge(new HashedPassword(MD5Crypt.crypt("foo"))).apply(challenge -> response));
   }
 
   @Test
   public void testAnonymousLoginBadName() {
-    final LobbyLoginValidator validator = new LobbyLoginValidator();
-    final SocketAddress address = new InetSocketAddress(5000);
     final String name = "bitCh" + Util.createUniqueTimeStamp();
-    final String mac = MacFinder.getHashedMacAddress();
     new BadWordController().addBadWord("bitCh");
-    final Map<String, String> properties = new HashMap<>();
-    properties.put(LobbyLoginValidator.ANONYMOUS_LOGIN, Boolean.TRUE.toString());
-    properties.put(LobbyLoginValidator.LOBBY_VERSION, LobbyServer.LOBBY_VERSION.toString());
-    assertEquals(LobbyLoginValidator.THATS_NOT_A_NICE_NAME, new LobbyLoginValidator()
-        .verifyConnection(validator.getChallengeProperties(name, address), properties, name, mac, address));
+    assertEquals(LobbyLoginValidator.THATS_NOT_A_NICE_NAME,
+        generateChallenge(name, new HashedPassword(MD5Crypt.crypt("foo"))).apply(challenge -> new HashMap<>(
+            Collections.singletonMap(LobbyLoginValidator.ANONYMOUS_LOGIN, Boolean.TRUE.toString()))));
+  }
+
+  @Test
+  public void testLegacyLogin() {
+    final String hashedPassword = MD5Crypt.crypt("foo");
+    final Map<String, String> response = new HashMap<>();
+    final ChallengeResultFunction challengeFunction = generateChallenge(new HashedPassword(hashedPassword));
+    assertNull(challengeFunction.apply(challenge -> {
+      response.put(LobbyLoginValidator.HASHED_PASSWORD_KEY, hashedPassword);
+      assertEquals(challenge.get(LobbyLoginValidator.SALT_KEY), MD5Crypt.getSalt(MD5Crypt.MAGIC, hashedPassword));
+      return response;
+    }));
+    // with a bad password
+    assertNotNull(challengeFunction.apply(challenge -> {
+      final Map<String, String> falsePassResponse = new HashMap<>(response);
+      falsePassResponse.put(LobbyLoginValidator.HASHED_PASSWORD_KEY, MD5Crypt.crypt("wrong"));
+      return falsePassResponse;
+    }));
+    // with a non existent user
+    assertNotNull(generateChallenge(null).apply(challenge -> response));
```

Comment:
This is more a comment for how we should be testing authentication in the future...

I was going to suggest we should be comparing the actual error message to the expected error message (or at least a portion of it) instead of just saying the error message should be non-`null` because these tests could be passing for the wrong reason (e.g. validation could fail because of a banned MAC address for some reason, but this test would continue to pass even if the password check logic were broken).  However, I realize in most tests in this fixture, the same error message (`"Invalid username or password"`) is returned to avoid disclosing any useful information to an attacker, so it would still be difficult to tell if a test is passing for the reason we think it is.

Maybe we should be returning an enum from these validation methods that specifies the exact cause of failure (wrong password, unknown user, etc.).  To avoid disclosure of the root cause of an authentication failure, each enum would have an associated end-user message that would be returned to a caller higher up in the stack at some point.  Then we can have different enum values for specific authentication failures, but they would all produce the same end-user message to keep an attacker guessing.